### PR TITLE
My code review changes

### PR DESCRIPTION
**Pull request by ikyled**
Pull request for code review changes

### CODE REVIEW

**Comment by ikyled** [comment]
**Analysis of the program**
The program is card management system that allows users to manage cards(like Pokémon cards). It has functionality for adding and searching for cards that show the market value and the popularity of the card. The unit test were straight forward and worked well making it easy to understand and the use of interfaces makes it pretty flexible. 

**Was the program available in UC Github on time?**
Yes, it was available on time.

**Is the program easy for you, as an outsider, to understand?**
Yes, the code was easy for me to understand, the names of the tests and other methods made it easy to know what the code was doing. 

**Does the program compile?**
Yes

**Rationale behind my changes**

- There wasn't a whole lot I saw wrong with the code or how it was setup originally so on improving efficiency and readability.
- I added a validation method so that validating is its own responsibility to make testing easier while also making the save method easier to read.
- for the toString, I added a StringBuilder for the method which makes the toString more efficient for methods that are called frequently.
- The last change I made was updating the logging statements to make it more of a detailed message to make it easier to resolve debug issues.

File: src/main/java/com/teamfive/disscard/DissCardController.java [hunk 1]
```diff
@@ -101,9 +101,10 @@ public ResponseEntity<Card> addCard(@RequestBody Card card) {
         HttpHeaders headers = new HttpHeaders();
         try {
             newCard = cardService.save(card);
+            log.info("Card added successfully: {}", newCard);
         }
         catch (Exception e) {
-            log.error(e.getMessage());
+            log.error("Error while adding card: {}. Exception: {}", card, e.getMessage());
```
Comment (ikyled) [inline src/main/java/com/teamfive/disscard/DissCardController.java:107]: Updated the logging statements to provide a more detailed message to make it easier to diagnose issues. Good logging practices lead to better observability to easily monitor performance.
https://www.slf4j.org/manual.html

File: src/main/java/com/teamfive/disscard/dto/Card.java [hunk 2]
```diff
@@ -39,7 +39,10 @@
     @SerializedName("popularity")
     public int popularity;
 
-    public String ToString() {
-        return cardName + " (" + series + ")";
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        sb.append(cardName).append(" (").append(series).append(")");
+        return sb.toString();
```
Comment (ikyled) [inline src/main/java/com/teamfive/disscard/dto/Card.java:46]: Added a string builder to toString to improve efficiency of the toString method to a method that is called often, which minimizes the overhead. 
https://docs.oracle.com/javase/8/docs/api/java/lang/StringBuilder.html

File: src/main/java/com/teamfive/disscard/service/CardServiceStub.java [hunk 3]
```diff
@@ -20,6 +21,10 @@ public CardServiceStub(ICardDAO cardDAO) {
         this.cardDAO = cardDAO;
     }
 
+    private boolean isValidCard(Card card) {
+        return card != null && card.getCardName() != null && !card.getCardName().isEmpty();
+    }
+
```
Comment (ikyled) [inline src/main/java/com/teamfive/disscard/service/CardServiceStub.java:27]: Introduced a validation method to make validating be a single responsibility, making the save method clearer and easier to test, improving readability, and simplifying future modifications.
https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html

**Comment by leichtje** [comment]
Confirmed Pull Request

**Pull request merged**